File: app/src/main/java/pe/andy/bookholic/searcher/impl/GangnamLibrarySearchTask.java
package pe.andy.bookholic.searcher.impl;

import org.apache.commons.lang3.RegExUtils;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.lang.ref.SoftReference;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.function.Function;
import java.util.stream.Collectors;

import lombok.Getter;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import pe.andy.bookholic.MainActivity;
import pe.andy.bookholic.model.Ebook;
import pe.andy.bookholic.model.SearchField;
import pe.andy.bookholic.model.SearchQuery;
import pe.andy.bookholic.searcher.LibrarySearchTask;
import pe.andy.bookholic.util.JsonParser;
import pe.andy.bookholic.util.SslTrust;
import pe.andy.bookholic.util.Str;

public class GangnamLibrarySearchTask extends LibrarySearchTask {

    @Getter final String libraryCode = "";
    static final String libraryName = "강남구 전자도서관";
    static final String baseUrl = "http://ebook.gangnam.go.kr";

    public GangnamLibrarySearchTask(MainActivity activity) {
        super(activity, libraryName, baseUrl);

        this.setEncoding(Encoding_EUCKR);
    }

    @Override
    public LibrarySearchTask create() {
        SoftReference<LibrarySearchTask> ref = new SoftReference<>(new GangnamLibrarySearchTask(this.mActivity));
        return ref.get();
    }

    @Override
    protected String getField(SearchQuery query) {
        Map<SearchField, String> fieldMap = new HashMap<>();
        fieldMap.put(SearchField.ALL, "도서명");
        fieldMap.put(SearchField.TITLE, "도서명");
        fieldMap.put(SearchField.AUTHOR, "저자명");
        fieldMap.put(SearchField.PUBLISHER, "출판사명");

        String field = fieldMap.get(query.getField());
        return field;
    }

    @Override
    protected Response request(SearchQuery query) throws IOException {

        String url = this.baseUrl + "/books/book_info.asp";
        String keyword = query.getEncodedKeyword("EUC-KR");

        HttpUrl.Builder builder = HttpUrl.parse(url).newBuilder()
                .addQueryParameter("page_num", query.getPageString())
                .addQueryParameter("list_num", "20")
                .addQueryParameter("ldav", "off")
                .addEncodedQueryParameter("bsc1", this.getEuckrString("도서검색"))
                .addEncodedQueryParameter("bsc2", this.getEuckrString(this.getField(query)))
                .addEncodedQueryParameter("sw", keyword);

        Request req = new Request.Builder()
                .url(builder.build().toString())
                .addHeader("Accept", "text/html; charset=euc-kr")
                .addHeader("User-Agent", userAgent)
                .build();

        OkHttpClient client = SslTrust.trustAllSslClient(new OkHttpClient());
        Response resp = client
                .newCall(req)
                .execute();

        return resp;
    }

    private String getEuckrString(String text) {
        try {
            return URLEncoder.encode(text, "EUC-KR");
        } catch (UnsupportedEncodingException e) {
            return "";
        }
    }

    @Override
    protected List<Ebook> parse(String html) throws IOException {
        Document doc = Jsoup.parse(html);
        Elements wrapper = doc.select(".book_list");

        // 전체 수량
        Elements elems = wrapper.select("p.location");
        elems.select("font").remove();
        this.resultCount = Str.extractInt(elems.text());

        // 페이지 수
        elems = wrapper.select(".paginate a");
        Element elem = elems.last();
        if (elem != null) {
            this.resultPageCount = Str.extractInt(elem.attr("href"));
        }
        else {
            this.resultPageCount = 1;
        }

        elems = wrapper.select("ul li .book");
        List<Ebook> list = elems.stream()
                .map(this.elementParser)
                .collect(Collectors.toList());
        return list;
    }

    private Function<Element, Ebook> elementParser = li -> {
        Ebook ebook = new Ebook(this.libraryName);

        Element elem = li.select("p.book_cover a").first();
        ebook.setUrl("https://ebook.gangnam.go.kr/books/" + elem.attr("href"));
        ebook.setThumbnailUrl(JsonParser.getAttrOfFirstElement(elem, "img", "src"));

        if (li.select("p.application span").size() > 1){
            elem = li.select("p.application span").get(1);
            ebook.setPlatform(elem.text());
        } else {
            ebook.setPlatform("");
        }

        ebook.setTitle(JsonParser.getTextOfFirstElement(li, "h1.title a"));
        ebook.setAuthor(JsonParser.getTextOfFirstElement(li, "h2.writer"));
        ebook.setPublisher(JsonParser.getTextOfFirstElement(li, "h3.publisher"));
        ebook.setDate(RegExUtils.replaceAll(JsonParser.getTextOfFirstElement(li, "h3.date"), "출판일", ""));

        Elements spans = li.select("p.state span.number");
        Queue<String> queue = new LinkedList<>(spans.eachText());
        ebook.setCountTotal(Str.extractInt(Str.def(queue.poll())));
        ebook.setCountRent(Str.extractInt(Str.def(queue.poll())));

        return ebook;
    };
}

File: app/src/main/java/pe/andy/bookholic/searcher/impl/GangnamLibrarySearchTask.kt
package pe.andy.bookholic.searcher.impl

import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import pe.andy.bookholic.fragment.SearchFragment
import pe.andy.bookholic.model.Ebook
import pe.andy.bookholic.model.Library
import pe.andy.bookholic.model.SearchField
import pe.andy.bookholic.model.SearchQuery
import pe.andy.bookholic.parser.GangnamLibraryParser
import pe.andy.bookholic.searcher.LibrarySearchTask
import pe.andy.bookholic.util.HttpExtension
import pe.andy.bookholic.util.SslTrust.Companion.trustAllSslClient
import pe.andy.bookholic.util.StringExtension
import java.io.IOException
import java.lang.ref.SoftReference

class GangnamLibrarySearchTask(
        searchFragment: SearchFragment
) : LibrarySearchTask(searchFragment, library),
        StringExtension, HttpExtension {

    init {
        this.encoding = library.encoding
    }

    companion object {
        val library = Library(name = "강남구 전자도서관", url = "https://ebook.gangnam.go.kr", encoding = Library.Encoding_EUCKR)
        val searchFields: Map<SearchField, String> = mapOf(
                SearchField.ALL to "도서명",
                SearchField.TITLE to "도서명",
                SearchField.AUTHOR to "저자명",
                SearchField.PUBLISHER to "출판사명"
        )
    }

    override fun getLibraryCode() = library.code

    override fun create(): LibrarySearchTask {
        return SoftReference<LibrarySearchTask>(
                GangnamLibrarySearchTask(searchFragment)
        ).get()!!
    }

    override fun getField(query: SearchQuery): String {
        return searchFields[query.field] ?: "도서명"
    }

    @Throws(IOException::class)
    override fun request(query: SearchQuery): Response {
        val req: Request = Request.Builder()
                .url(getUrl(query))
                .accept(Encoding_EUCKR)
                .userAgent(userAgent)
                .build()

        return trustAllSslClient(OkHttpClient())
                .newCall(req)
                .execute()
    }

    private fun getUrl(query: SearchQuery): String {
        val url = "${library.url}/books/book_info.asp"

        return url.toHttpUrlOrNull()!!.newBuilder()
            .addQueryParameters(
                mapOf(
                    "page_num" to query.page.toString(),
                    "list_num" to "20",
                    "ldav" to "off"
                )
            )
                .addEncodedQueryParameters(mapOf(
                        "bsc1" to "도서검색".encodeToEucKR(),
                        "bsc2" to getField(query).encodeToEucKR(),
                        "sw" to query.keyword.encodeToEucKR()
                ))
                .build()
                .toString()
    }

    @Throws(IOException::class)
    override fun parse(html: String): List<Ebook> {
        val doc: Document = Jsoup.parse(html)

        parseMetaCount(doc)
        return parseBook(doc)
    }

    private fun parseMetaCount(doc: Document) {
        val pair = GangnamLibraryParser.parseMetaCount(doc, library)

        resultPageCount = pair.second
        resultCount = pair.first

        // 강남구 전자도서관에서 검색결과 개수를 알려주지 않음
        if (resultCount == -1 && resultPageCount > 0) {
            resultCount = resultPageCount * 20
        }
    }

    private fun parseBook(doc: Document): List<Ebook> {
        if (resultCount <= 0) {
            return emptyList()
        }

        return doc.select(".book_list .book")
                .asSequence()
                .map { GangnamLibraryParser.parse(it, library = library) }
                .toList()
    }
}